### PR TITLE
feat: update multiqc to 1.22.3

### PR DESCRIPTION
**Pull request by mathiasbio**
## Description

The new version of multiqc supports picard mimicked reports from Sentieon tools: https://github.com/MultiQC/MultiQC/pull/2110

This should solve this issue: https://github.com/Clinical-Genomics/BALSAMIC/issues/1290 where an ugly solution was implemented in the Dedup rule to make MultiQC accept dedup-stats from Sentieon dedup. 

It may also allow us to move away from Picard to generate our QC reports and instead use the Sentieon tools which should be faster and enable us to clear away some rules for a more streamlined and less messy workflow. 

#### Added

- separate container for multiqc

#### Changed

- updated multiqc from 1.12 to 1.22.3

#### Removed

- no longer necessary sed command in dedup rule 
- deprecated and unused TNhaplotyper rule

## Documentation

- [ ] N/A
- [x] Updated Balsamic documentation to reflect the changes as needed for this PR.
    - read the docs: bioinfo_softwares

## Tests

`
I downloaded the new multiqc docker (v1.22.3 of multiqc) as a singularity image and placed in a separate develop cache which I then used in my tests along with the updated align_qc.sif container which doesn't contain multiqc anymore `

**Ensure that we get all QC metrics in deliverables as before**

**NOTE:** regarding multiqc version

- As of 2024-04-31 it seems that there's a bug in multiqc or a feature, I don't know, but the latest version 1.22.1 does not pick up the relatedness score from somalier which kind of blocks this feature at the moment. But I have opened up a bug report in multiqc repo: https://github.com/MultiQC/MultiQC/issues/2596 and have tried downgrading the version to 1.18 which seems to be working. 
- As of 2024-06-05 They solved the bug and I have tested the latest version 1.22.2 which does pick up Somalier results, but I discovered that it also does not manage to pick up the Picard WgsMetrics results, which v1.18 does. So I will return to that older version. I opened a bug report https://github.com/MultiQC/MultiQC/issues/2607
- As of 2024-06-22 they are supposed to have solved the bug with version **1.22.3**

- [x] Verified successfully for WGS T+N, and TGA-UMI T+N all metrics in the deliverables file are still there as in version 15.0.0

### Feature Tests

- [x] Unaltered dedup metrics file from Sentieon dedup is picked up by multiqc. **However!** the name of the tool in multiqc is still referred to as Picard, not Sentieon. But at least the raw metric file is unaltered and can be traced to Sentieon. 

### Pipeline Integrity Tests

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
    - [ ] Verified
- **TGA T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **TGA T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/N Workflow**
    - [ ] N/A
    - [x] Verified
- **WGS T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/N Workflow**
    - [ ] N/A
    - [x] Verified
- **QC Workflow**
    - [x] N/A
    - [ ] Verified
- **PON Workflow**
    - [x] N/A
    - [ ] Verified

## Clinical Genomics Stockholm

<details>
<summary></summary>

### Documentation

- **Atlas documentation**
    - [x] N/A
    - [ ] Updated: [Link]
- **Web portal for Clinical Genomics**
    - [x] N/A
    - [ ] Updated: [Link]

### User Changes

- [x] N/A
- [ ] This PR affects the output files or results.
    - [ ] User feedback is considered unnecessary because [Justification].
    - [ ] Affected users have been included in the development process and given a chance to provide feedback.

### Infrastructure Changes

- **Stored files in Housekeeper**
    - [x] N/A
    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Servers (configuration files on Hasta)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Scout interface**
    - [x] N/A
    - [ ] Updated: [Link]

</details>

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [x] Provided a comprehensive description of the PR.
    - [x] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [x] Verified and updated documentation if necessary.
- **Tests**
    - [x] Described and tested the functionality addressed in the PR.
    - [x] Ensured integration of the new code with existing workflows.
    - [x] Confirmed that meaningful unit tests were added for the changes introduced.
    - [x] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [ ] Addressed and resolved all the feedback provided during the code review process.
    - [ ] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [x] Code implements the intended features or fixes the reported issue.
    - [x] Code follows the project's coding standards and style guide.
- **Documentation**
    - [x] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Tests**
    - [x] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [x] Confirmed that the developer has addressed all the comments during the code review.

### CODE REVIEW

File: BALSAMIC/assets/scripts/collect_qc_metrics.py [hunk 1]
```diff
@@ -186,41 +186,49 @@ def get_metric_condition(
     return req_metrics
 
 
+def get_sample_id(multiqc_key: str):
+    """Returns extracted sample id from multiqc data json key"""
+    # example of possible sample-formats below from "report_saved_raw_data":
+    # tumor.ACCXXXXXX
+    # tumor.ACCXXXXXX_FR
+    # ACCXXXXXX_align_sort_HMYLNDSXX_ACCXXXXXX_S165_L001
+    # extracted below for id to: ACCXXXXXX
```
Comment (ivadym) [inline BALSAMIC/assets/scripts/collect_qc_metrics.py:195]: I would include these comments in the docstring

File: BALSAMIC/containers/multiqc/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,30 @@
+FROM python:3.11.3-slim
+
+LABEL base.image="python:3.11.3-slim"
+LABEL about.home="https://github.com/Clinical-Genomics/BALSAMIC"
+LABEL about.documentation="https://balsamic.readthedocs.io/"
+LABEL about.license="MIT License (MIT)"
+LABEL about.description="Bioinformatic analysis pipeline for somatic mutations in cancer"
```
Comment (ivadym) [inline BALSAMIC/containers/multiqc/Dockerfile:7]: LABEL base.image="python:3.11.3-slim" \
    about.home="https://github.com/Clinical-Genomics/BALSAMIC" \
    about.documentation="https://balsamic.readthedocs.io/" \
    about.license="MIT License (MIT)" \
    about.description="Bioinformatic analysis pipeline for somatic mutations in cancer"

Comment (ivadym) [reply BALSAMIC/containers/multiqc/Dockerfile:7]: And this should be a description of the container itslef, not Balsamic. I think we have some examples in other containers

File: BALSAMIC/containers/multiqc/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+# Install pip and multiqc, create user and home directory
+RUN python -m pip install --upgrade pip && \
+    pip install multiqc==1.22.3 && \
+    adduser --disabled-password --gecos '' "$USER" && \
+    mkdir -p "$HOME" && \
+    chown -R "$USER:$USER" "$HOME"
+
+# Switch to the non-root user
+USER "$USER"
+WORKDIR "$HOME"
```
Comment (ivadym) [inline BALSAMIC/containers/multiqc/Dockerfile:24]: `WORKDIR "$HOME"` directive also creates the folder if it doesn't exit. So you should move this part before the RUN command and remove the `mkdir`

**Review by ivadym — COMMENTED** [review]
Fantastic work ⭐

File: BALSAMIC/assets/scripts/collect_qc_metrics.py [hunk 4]
```diff
@@ -186,41 +186,49 @@ def get_metric_condition(
     return req_metrics
 
 
+def get_sample_id(multiqc_key: str):
```
[the remark this answers is not in the document]
Comment (mathiasbio) [reply BALSAMIC/assets/scripts/collect_qc_metrics.py:189]: Done!

File: BALSAMIC/containers/multiqc/Dockerfile (hunk 3 above)
Comment (mathiasbio) [reply BALSAMIC/containers/multiqc/Dockerfile]: Done!

File: BALSAMIC/containers/multiqc/Dockerfile (hunk 2 above)
Comment (mathiasbio) [reply BALSAMIC/containers/multiqc/Dockerfile]: Nice! Done!

File: BALSAMIC/containers/align_qc/align_qc.yaml [hunk 5]
```diff
@@ -122,7 +122,6 @@ dependencies:
   - markdown=3.4.1
   - markupsafe=2.1.1
   - matplotlib-base=3.6.2
-  - multiqc=1.12
```
[the remark this answers is not in the document]
Comment (ivadym) [reply BALSAMIC/containers/align_qc/align_qc.yaml:125]: Not sure it's worth it though, I'm fine if you leave it as it is and it works. We could have an independent PR updating our containts

File: BALSAMIC/assets/scripts/collect_qc_metrics.py [hunk 6]
```diff
@@ -186,41 +186,53 @@ def get_metric_condition(
...
```
Comment (ivadym) [inline BALSAMIC/assets/scripts/collect_qc_metrics.py:189]:
```suggestion
def get_sample_id(multiqc_key: str) -> str:
```

File: BALSAMIC/assets/scripts/collect_qc_metrics.py [hunk 7]
```diff
@@ -186,41 +186,53 @@ def get_metric_condition(
...
+    """Returns extracted sample ID from MultiQC data JSON key.
+
+    Example of possible sample-formats below from "report_saved_raw_data":
+    tumor.ACCXXXXXX
+    tumor.ACCXXXXXX_FR
+    ACCXXXXXX_align_sort_HMYLNDSXX_ACCXXXXXX_S165_L001
+
+    returns ACCXXXXXX
```
Comment (ivadym) [inline BALSAMIC/assets/scripts/collect_qc_metrics.py:197]:
```suggestion
    Returns:
        str: The extracted sample ID with the ACCXXXXXX format.
```

File: CHANGELOG.rst [hunk 8]
```diff
@@ -10,6 +10,7 @@ Changed:
 ^^^^^^^^
 * Cluster scheduler script for immediate submit https://github.com/Clinical-Genomics/BALSAMIC/pull/1372
 * `SLEEP_BEFORE_START` to 600s https://github.com/Clinical-Genomics/BALSAMIC/pull/1372
+* Updated multiqc to version 1.22.3 https://github.com/Clinical-Genomics/BALSAMIC/pull/1441
```
Comment (ivadym) [inline CHANGELOG.rst:13]:
```suggestion
* Updated MultiQC to version 1.22.3 https://github.com/Clinical-Genomics/BALSAMIC/pull/1441
```

File: tests/scripts/test_collect_qc_metrics.py [hunk 9]
```diff
@@ -284,6 +285,17 @@ def test_collect_qc_metrics_counts(
     assert Path(output_path).exists()
 
 
+def test_get_sample_id():
+    """Tests sample id extraction from multiqc_key"""
```
Comment (ivadym) [inline tests/scripts/test_collect_qc_metrics.py:289]:
```suggestion
    """Tests sample ID extraction from multiqc_key."""
```

File: tests/scripts/test_collect_qc_metrics.py [hunk 10]
```diff
@@ -284,6 +285,17 @@ def test_collect_qc_metrics_counts(
...
+    multiqc_sampleid_keys = [
+        "tumor.sampleid",
+        "tumor.sampleid_R1",
+        "sampleid_align_sort_HMYLNDSXX_sampleid_S165_L001",
+    ]
+    for multiqc_key in multiqc_sampleid_keys:
+        assert get_sample_id(multiqc_key) == "sampleid"
```
Comment (ivadym) [inline tests/scripts/test_collect_qc_metrics.py:296]:
```suggestion
def test_get_sample_id():
    """Tests sample id extraction from multiqc_key"""
    multiqc_sample_ids = [
        "tumor.{tumor_sample_name}",
        "tumor.{tumor_sample_name}_R1",
        f"{tumor_sample_name}_align_sort_HMYLNDSXX_{tumor_sample_name}_S165_L001",
    ]
    for multiqc_key in multiqc_sample_ids:
        assert get_sample_id(multiqc_key) == tumor_sample_name
```

**Review by ivadym — APPROVED** [review]
💯

File: tests/scripts/test_collect_qc_metrics.py (hunk 10 above)
Comment (mathiasbio) [reply tests/scripts/test_collect_qc_metrics.py:296]: Nice idea, done!

**Comment by mathiasbio** [comment]
Thanks a lot for the review @ivadym ! ❤️

**Comment by mathiasbio** [comment]
I will pull down the multiqc container to confirm it's still working, then squash and merge
